src/charm.py

```diff
@@ -0,0 +1,477 @@
+            event.defer()
+        except ops.pebble.ChangeError as e:
+            self.unit.status = ops.BlockedStatus(f"Could not start {container.name}: {e}")
+            event.defer()
+
+    def _replan(self, container: ops.Container):
+        summary, command, update_version = self.container_details[container.name]
+        logger.debug("Updating plan for %s (%s -> %s)", container.name, summary, command)
+        layer = ops.pebble.Layer(
+            {
+                "summary": summary,
+                "description": "Service for Takahē ActivityPub instance",
+                "services": {
+                    container.name: {
+                        "override": "replace",
+                        "summary": summary,
+                        "command": command,
+                        "startup": "enabled",
+                        "environment": self._takahē_env,
+                    },
+                },
+            }
+        )
+        new_layer = layer.to_dict()
+        services = container.get_plan().to_dict().get("services", {})
+        if services != new_layer.get("services"):
+            container.add_layer(container.name, layer, combine=True)
+            logger.info("Added updated layer %r to Pebble plan.", container.name)
+            container.replan()
+            logger.info("Updated services.")
+
+        if update_version:
+            self.unit.set_workload_version(self.workload_version)
+
+    @property
+    def workload_version(self):
+        """The version of Takahē installed in the containers."""
+        logger.debug("Getting the workload version.")
+        # This can be done on either container, so use the background service
+        # to minimise the impact.
+        container = self.unit.get_container(self.background_service_name)
+        try:
+            python = container.exec(
+                ["python", "-c", "import takahe;print(takahe.__version__)"], working_dir="/takahe"
+            )
+            return python.wait_output()[0]
+        except (ops.pebble.APIError, ops.pebble.ExecError) as e:
+            logger.warning("Unable to get version from Takahē: %s", e)
+            return "unknown"
+
+    @property
+    def dsn(self):
+        """The string used to connect to the related PostgreSQL database."""
+        if "database" not in self.model.relations or not self.model.relations["database"]:
+            logger.debug("Cannot get DSN: relation is not ready.")
+            return ""
+        relation = self.model.relations["database"][0]
+        # TODO: Is there some better way to do this? The secrets are provided as
+        # attributes of the database events, but we need to provide them in the
+        # environment every time we replan(). We don't really want to copy them
+        # into a local secret. It seems like the interface sets a label, so we
+        # can't just access them that way, unless we get the label, but that's
+        # also hidden behind a private method.
```

Comment:
Hmm, yeah, it seems less than ideal that the charm lib can't provide this when you ask for it. Maybe talk to the charm lib authors to see if they have any better ideas here?